### DISCUSSION TITLE
detailed information on CRUD operations in DID

### DISCUSSION THREAD

**Discussion by praveensankar** [decentralized-identity/veramo#392] [Q&A]
I am looking for the flow for CRUD operations in DID in the architecture level. I want to know information related to which components are involved in DID creation and how it is added in any of the live test / main network's storage. Is there any documentation available for this ?

**Comment by mirceanis** [comment]
At the moment, the veramo framework can be used with DID providers for `did:ethr`, `did:web`, and `did:key` for CRUD operations.
None of these DID methods require on-chain storage during creation.
`did:web` and `did:key` have no blockchain component, and `did:ethr`, by design, does not require working with an ethereum network for _creation_, only for _update_, _read_ and _delete_.

Please read my answer to a similar question about `did:ethr`: https://github.com/uport-project/veramo/discussions/365#discussioncomment-335167. That answer links to the relevant documentation for `did:ethr` if you wish to find out more.

---

Please note that other DID methods do use a network during creation, but there is no plugin that we know of that implements them for `veramo` yet.
For such an implementation to happen, one would need to create an implementation of [AbstractIdentifierProvider](https://github.com/uport-project/veramo/blob/650d73c04e83ea4b8a10f05578bd00f10c579538/packages/did-manager/src/abstract-identifier-provider.ts#L7) which would then be used in the to the constructor of the [DIDManager](https://github.com/uport-project/veramo/blob/650d73c04e83ea4b8a10f05578bd00f10c579538/packages/did-manager/src/id-manager.ts#L27) plugin (see example [here](https://github.com/uport-project/veramo/blob/650d73c04e83ea4b8a10f05578bd00f10c579538/__tests__/localAgent.test.ts#L104)). Then, with that plugin in use, one would call `agent.didManagerCreate(...)` (see example [here](https://github.com/uport-project/veramo/blob/650d73c04e83ea4b8a10f05578bd00f10c579538/__tests__/shared/didManager.ts#L27))
